File: 146-lru-cache/lru-cache.cpp
class LRUCache {
public:

    struct ListNode{
        int key;
        int val;
        ListNode* left;
        ListNode* right;
        ListNode(int key, int val, ListNode *left, ListNode* right){
            this->key = key;
            this->val = val;
            this->left = left;
            this->right = right;
        }
    };

    int maxCap;
    ListNode* startNode;
    ListNode* endNode;
    unordered_map<int,ListNode*> mp;

    LRUCache(int capacity) {
        maxCap = capacity;
        startNode = new ListNode(0,0,NULL,NULL);
        endNode = new ListNode(0,0,NULL,NULL);
        startNode->right = endNode;
        endNode->left = startNode;
    }
    
    void insertNodeAtStart(ListNode* root){
        ListNode* prevNode = startNode;;
        ListNode* nextNode = startNode->right;
        
        prevNode->right = root;
        nextNode->left = root;

        root->right = nextNode;
        root->left = prevNode;
    }

    void removeNode(ListNode* root){
        ListNode* rootleft = root->left;
        ListNode* rootright = root->right;
        rootleft->right = rootright;
        rootright->left = rootleft;
    }

    int get(int key) {
        if(mp.find(key)==mp.end()){
            return -1;
        }
        ListNode* root = mp[key];
        removeNode(root);
        insertNodeAtStart(root);
        return root->val;
    }
    
    void put(int key, int value) {
        if(mp.find(key)!=mp.end()){
            mp[key]->val = value;
            removeNode(mp[key]);
            insertNodeAtStart(mp[key]);
            return;
        }
        if(mp.size()==maxCap){
            mp.erase(endNode->left->key);
            removeNode(endNode->left);
        }
        mp[key] = new ListNode(key,value,NULL,NULL);
        insertNodeAtStart(mp[key]);
        return;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146-lru-cache/lru-cache.py
class ListNode:

    def __init__(self, key: int, val: int , prev=None, next=None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.left_node = ListNode(0, 0, None, None)
        self.right_node = ListNode(0, 0,self.left_node, None)
        self.left_node.next = self.right_node
        self.map = {}

    def remove(self, cur_node: ListNode):
            prev_node = cur_node.prev
            next_node = cur_node.next
            prev_node.next = next_node
            next_node.prev = prev_node

    def add_to_front(self, cur_node: ListNode):
        prevnode = self.left_node
        nextnode = self.left_node.next
        
        cur_node.prev = prevnode
        cur_node.next = nextnode

        prevnode.next = cur_node
        nextnode.prev = cur_node
    
    
    def get(self, key: int) -> int:
        
        if key in self.map:
            node = self.map[key]
            self.remove(node)
            self.add_to_front(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        
        if(self.get(key) != -1):
            node = self.map[key]
            node.val = value
            return

        if(len(self.map)==self.capacity):
            node = self.right_node.prev
            self.remove(node)
            del self.map[node.key]
            
        
        node = ListNode(key,value)
        self.add_to_front(node)
        self.map[key] = node


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)